File: Polimorfismo - Batalha Final/Item.java
/* USO DE IA */
/* Ferramenta utilizada: ChatGPT */
/* A IA foi utilizada para auxiliar na implementação da classe abstrata, polimorfismo, 
método habilidade() e no uso de instanceof.*/
/* O instanceof foi usado apenas para identificar os Magos e acessar uma característica exclusiva deles: a mana, não 
foi usado dentro do laço da batalha porque isso faria verificações do tipo de cada personagem. */

import java.util.ArrayList;
import java.util.List;

class Item {

    private String nome;
    private int bonus;

    public Item(String nome, int bonus) {
        this.nome = nome;
        setBonus(bonus);
    }

    public String getNome() {
        return nome;
    }

    public int getBonus() {
        return bonus;
    }

    public void setBonus(int bonus) {
        if (bonus >= 0) {
            this.bonus = bonus;
        } else {
            System.out.println("Bonus não pode ser negativo.");
        }
    }

    public String descricao() {
        return nome + " (+" + bonus + ")";
    }
}

abstract class Personagem {

    protected String nome;
    protected int vida;
    protected int nivel;
    protected int forca;

    private Item[] inventario;
    private int quantidadeItens;

    public Personagem(String nome, int vida, int nivel, int forca) {
        setNome(nome);
        setVida(vida);
        setNivel(nivel);
        setForca(forca);

        inventario = new Item[10];
        quantidadeItens = 0;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        if (!nome.equals("")) {
            this.nome = nome;
        } else {
            System.out.println("Nome não pode ser vazio.");
        }
    }

    public int getVida() {
        return vida;
    }

    public void setVida(int vida) {
        if (vida >= 0 && vida <= 200) {
            this.vida = vida;
        } else {
            System.out.println("Vida deve estar entre 0 e 200.");
        }
    }

    public int getNivel() {
        return nivel;
    }

    public void setNivel(int nivel) {
        if (nivel >= 1) {
            this.nivel = nivel;
        } else {
            System.out.println("Nivel deve ser maior ou igual a 1.");
        }
    }

    public int getForca() {
        return forca;
    }

    public void setForca(int forca) {
        if (forca >= 0) {
            this.forca = forca;
        } else {
            System.out.println("Forca não pode ser negativa.");
        }
    }

    public void pegar(Item item) {
        if (quantidadeItens < inventario.length) {
            inventario[quantidadeItens] = item;
            quantidadeItens++;
        } else {
            System.out.println("Inventario cheio.");
        }
    }

    public void receberDano(int dano) {
        setVida(Math.max(0, vida - dano));
    }

    public void atacar(Personagem alvo) {
        alvo.receberDano(forca);
    }

    public boolean estaVivo() {
        return vida > 0;
    }

    public String ficha() {

        String resultado =
                "Nome: " + nome +
                "\nVida: " + vida +
                "\nNivel: " + nivel +
                "\nForca: " + forca +
                "\nInventario:";

        if (quantidadeItens == 0) {
            resultado += "\nNenhum item";
        } else {
            for (int i = 0; i < quantidadeItens; i++) {
                resultado += "\n- " + inventario[i].descricao();
            }
        }

        return resultado;
    }

    public abstract String habilidade();
}

class Mago extends Personagem {

    private int mana;

    public Mago(String nome, int vida, int nivel, int forca) {
        super(nome, vida, nivel, forca);
        mana = 50;
    }

    public int getMana() {
        return mana;
    }

    public void setMana(int mana) {
        if (mana >= 0) {
            this.mana = mana;
        } else {
            System.out.println("Mana não pode ser negativa.");
        }
    }

    @Override
    public String habilidade() {
        return "rajada arcana";
    }

    @Override
    public String ficha() {
        return super.ficha() +
                "\nMana: " + mana;
    }
}

class Guerreiro extends Personagem {

    private int defesa;

    public Guerreiro(String nome, int vida, int nivel, int forca) {
        super(nome, vida, nivel, forca);
        defesa = 5;
    }

    public int getDefesa() {
        return defesa;
    }

    public void setDefesa(int defesa) {
        if (defesa >= 0) {
            this.defesa = defesa;
        } else {
            System.out.println("Defesa não pode ser negativa.");
        }
    }

    @Override
    public void receberDano(int dano) {
        int danoEfetivo = Math.max(0, dano - defesa);
        super.receberDano(danoEfetivo);
    }

    @Override
    public String habilidade() {
        return "escudo de aço";
    }

    @Override
    public String ficha() {
        return super.ficha() +
                "\nDefesa: " + defesa;
    }
}

class Chefe extends Personagem {

    public Chefe(String nome) {
        super(nome, 200, 1, 20);
    }

    @Override
    public String habilidade() {
        return "ataque devastador";
    }

    @Override
    public String ficha() {
        return "[CHEFE] " + nome +
                " (vida: " + vida +
                ", forca: " + forca + ")";
    }
}

public class Main {

    public static void main(String[] args) {

        Item espada = new Item("Espada de Aço", 3);

        Mago mago = new Mago("Elara", 100, 1, 30);
        mago.pegar(espada);

        Guerreiro guerreiro = new Guerreiro("Bran", 100, 1, 25);

        Chefe chefe = new Chefe("Rei Lich");

        List<Personagem> herois = new ArrayList<>();

        herois.add(mago);
        herois.add(guerreiro);

        System.out.println("----- BATALHA FINAL -----");
        System.out.println();

        for (Personagem heroi : herois) {

            System.out.println(heroi.ficha());

            System.out.println(
                    heroi.getNome() +
                    " usa " +
                    heroi.habilidade()
            );

            heroi.atacar(chefe);

            System.out.println(
                    "Rei Lich recebeu " +
                    heroi.getForca() +
                    " de dano."
            );

            System.out.println();
        }


        System.out.println("----- RESULTADO -----");
        System.out.println(chefe.ficha());

        for (Personagem personagem : herois) {

            if (personagem instanceof Mago) {

                Mago magoAtual = (Mago) personagem;

                System.out.println(
                        "Mana do Mago " +
                        magoAtual.getNome() +
                        ": " +
                        magoAtual.getMana()
                );
            }
        }
    }
}

File: Polimorfismo - Batalha Final/Item.py
/* USO DE IA */

/* Ferramenta utilizada: ChatGPT */
/*A IA foi utilizada para auxiliar na aplicação da classe abstrata, polimorfismo, método habilidade () e o uso do isistance*/
/* O isinstance foi usado apenas para identificar os Magos e acessar uma característica exclusiva deles: a mana. Não foi usado
isinstance dentro do laço da batalha porque isso faria verificações do tipo de cada personagem.*/

from abc import ABC, abstractmethod

class Item:

    def __init__(self, nome, bonus):
        self.__nome = nome
        self.set_bonus(bonus)

    def get_nome(self):
        return self.__nome

    def get_bonus(self):
        return self.__bonus

    def set_bonus(self, bonus):
        if bonus >= 0:
            self.__bonus = bonus
        else:
            print("Bonus não pode ser negativo.")

    def descricao(self):
        return self.__nome + " (+" + str(self.__bonus) + ")"

class Personagem(ABC):

    def __init__(self, nome, vida, nivel, forca):
        self.set_nome(nome)
        self.set_vida(vida)
        self.set_nivel(nivel)
        self.set_forca(forca)

        self.__inventario = [None] * 10
        self.__quantidadeItens = 0

    def get_nome(self):
        return self.__nome

    def set_nome(self, nome):
        if nome != "":
            self.__nome = nome
        else:
            print("Nome não pode ser vazio.")

    def get_vida(self):
        return self.__vida

    def set_vida(self, vida):
        if 0 <= vida <= 200:
            self.__vida = vida
        else:
            print("Vida deve estar entre 0 e 200.")

    def get_nivel(self):
        return self.__nivel

    def set_nivel(self, nivel):
        if nivel >= 1:
            self.__nivel = nivel
        else:
            print("Nivel deve ser maior ou igual a 1.")

    def get_forca(self):
        return self.__forca

    def set_forca(self, forca):
        if forca >= 0:
            self.__forca = forca
        else:
            print("Forca não pode ser negativa.")

    def pegar(self, item):
        if self.__quantidadeItens < len(self.__inventario):
            self.__inventario[self.__quantidadeItens] = item
            self.__quantidadeItens += 1
        else:
            print("Inventario cheio.")

    def receber_dano(self, dano):
        self.set_vida(max(0, self.__vida - dano))

    def atacar(self, alvo):
        alvo.receber_dano(self.__forca)

    def ficha(self):

        resultado = (
            "Nome: " + self.__nome +
            "\nVida: " + str(self.__vida) +
            "\nNivel: " + str(self.__nivel) +
            "\nForca: " + str(self.__forca) +
            "\nInventario:"
        )

        if self.__quantidadeItens == 0:
            resultado += "\nNenhum item"
        else:
            for i in range(self.__quantidadeItens):
                resultado += "\n- " + self.__inventario[i].descricao()

        return resultado

    @abstractmethod
    def habilidade(self):
        pass


class Mago(Personagem):

    def __init__(self, nome, vida, nivel, forca):
        super().__init__(nome, vida, nivel, forca)
        self.__mana = 50

    def get_mana(self):
        return self.__mana

    def set_mana(self, mana):
        if mana >= 0:
            self.__mana = mana
        else:
            print("Mana não pode ser negativa.")

    def habilidade(self):
        return "rajada arcana"

    def ficha(self):
        return super().ficha() + "\nMana: " + str(self.__mana)


class Guerreiro(Personagem):

    def __init__(self, nome, vida, nivel, forca):
        super().__init__(nome, vida, nivel, forca)
        self.__defesa = 5

    def get_defesa(self):
        return self.__defesa

    def set_defesa(self, defesa):
        if defesa >= 0:
            self.__defesa = defesa
        else:
            print("Defesa não pode ser negativa.")

    def receber_dano(self, dano):
        dano_efetivo = max(0, dano - self.__defesa)
        super().receber_dano(dano_efetivo)

    def habilidade(self):
        return "escudo de aço"

    def ficha(self):
        return super().ficha() + "\nDefesa: " + str(self.__defesa)


class Chefe(Personagem):

    def __init__(self, nome):
        super().__init__(nome, 200, 1, 20)

    def habilidade(self):
        return "ataque devastador"

    def ficha(self):
        return (
            "[CHEFE] " + self.get_nome() +
            " (vida: " + str(self.get_vida()) +
            ", forca: " + str(self.get_forca()) + ")"
        )

espada = Item("Espada de Aço", 3)

mago = Mago("Elara", 100, 1, 30)
mago.pegar(espada)

guerreiro = Guerreiro("Bran", 100, 1, 25)

chefe = Chefe("Rei Lich")

herois = [mago, guerreiro]

print("----- BATALHA FINAL -----")
print()

for heroi in herois:

    print(heroi.ficha())
    print(heroi.get_nome() + " usa " + heroi.habilidade())

    heroi.atacar(chefe)

    print("Rei Lich recebeu",
          heroi.get_forca(), "de dano.")
    print()


print("----- RESULTADO -----")
print(chefe.ficha())

for personagem in herois:

    if isinstance(personagem, Mago):
        print(
            "Mana do Mago",
            personagem.get_nome() + ":",
            personagem.get_mana()
        )
